File: dart/words/test/words_test.dart
import 'package:words/words.dart';
import 'package:test/test.dart';

void main() {
  group("Words", (){
    test('Reverse String', () {
      expect(Words.reverseString('hello'), equals('olleh'));
      expect(Words.reverseString('a'), equals('a'));
      expect(Words.reverseString(''), equals(''));
    });

    test('Is Palindrome', () {
      expect(Words.isPalindrome('race car'), equals(true));
      expect(Words.isPalindrome('level'), equals(true));
      expect(Words.isPalindrome('hello'), equals(false));
      expect(Words.isPalindrome('a'), equals(true));
      expect(Words.isPalindrome(''), equals(true));
    });

    test('Is Substring', () {
      expect(Words.kmpSearch('test', 'this is a test'), equals(true));
      expect(Words.kmpSearch('not', 'this is a test'), equals(false));
      expect(Words.kmpSearch('', 'any string'), equals(true));
      expect(Words.kmpSearch('abc', 'abc'), equals(true));
      expect(Words.kmpSearch('abc', 'ab'), equals(false));
    });

    test('Largest Common Subsequence', () {
      expect(Words.lcs('AGGTAB', 'GXTXAYB'), equals(4));
      expect(Words.lcs('ABC', 'ABC'), equals(3));
      expect(Words.lcs('ABC', 'DEF'), equals(0));
      expect(Words.lcs('', 'ABC'), equals(0));
      expect(Words.lcs('ABC', ''), equals(0));
    });
  });
}

File: zig/words/test/words_test.zig
const std = @import("std");
const words = @import("words");

test "reverse string" {
    try std.testing.expect(std.mem.eql(u8, words.reverse_string("hello"), "olleh"));
    try std.testing.expect(std.mem.eql(u8, words.reverse_string("a"), "a"));
    try std.testing.expect(std.mem.eql(u8, words.reverse_string(""), ""));
}

test "is palindrome" {
    try std.testing.expect(words.is_palindrome("race car") catch false);
    try std.testing.expect(words.is_palindrome("level") catch false);
    try std.testing.expect(!(try words.is_palindrome("hello")));
    try std.testing.expect(words.is_palindrome("a") catch false);
    try std.testing.expect(words.is_palindrome("") catch false);
}

test "is substring" {
    try std.testing.expect(words.kmp_search("test", "this is a test") catch false);
    try std.testing.expect(!(try words.kmp_search("not", "this is a test")));
    try std.testing.expect(words.kmp_search("", "any string") catch false);
    try std.testing.expect(words.kmp_search("abc", "abc") catch false);
    try std.testing.expect(!(try words.kmp_search("abc", "ab")));
}

test "longest common subsequence" {
    try std.testing.expect((try words.lcs("AGGTAB", "GXTXAYB")) == 4);
    try std.testing.expect((try words.lcs("ABC", "ABC")) == 3);
    try std.testing.expect((try words.lcs("ABC", "DEF")) == 0);
    try std.testing.expect((try words.lcs("", "ABC")) == 0);
    try std.testing.expect((try words.lcs("ABC", "")) == 0);
}
